File: 프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.js
function solution(ingredient) {
    
    let count = 0;
    
    let hamberger = [];
    
    for(let i = 0; i < ingredient.length; i++){
        const source = ingredient[i];
        
        if(source === 1){
            // 마지막 요소가 3인 경우 개수 추가 
            if(hamberger[hamberger.length - 3] === 1 && hamberger[hamberger.length - 2] === 2 && hamberger[hamberger.length - 1] === 3){
                count++;
                // 기존 1, 2, 3 지우기 
                hamberger.pop();
                hamberger.pop();
                hamberger.pop();
            } else {
                hamberger.push(source);
            }
        } else {
            hamberger.push(source);
        }
    }
    return count;
}

File: 프로그래머스/1/133502. 햄버거 만들기/햄버거 만들기.py
def solution(ingredient):
    stack = []
    count = 0
    
    for igdt in ingredient:
        
        # stack에 저장된 개수가 3개 이상인 경우에만 분기
        if len(stack) > 2:
            # 현재 재료가 1인 경우
            if igdt == 1:
                # 스택의 마지막 재료가 3인 경우
                last = stack.pop()
                if last == 3:
                    sec_last = stack.pop()
                    if sec_last == 2:
                        third_last = stack.pop()
                        if third_last == 1:
                            count += 1
                            continue
                        else:
                            stack.append(third_last)
                            stack.append(sec_last)
                            stack.append(last)
                    else:
                        stack.append(sec_last)
                        stack.append(last)
                else:
                    stack.append(last)
        
        stack.append(igdt)
        
    # print(stack)
    return count